### PR TITLE
samples: add schema samples, typescript sample support

### LINKED ISSUE #1220: Add support for Pub/Sub server schema to the Node client library
This feature is currently in preview, and we need to get support for it in the library. Some things that need to be included (may expand):

* Support in the GAPIC layer (done)
* Support in the veneer (PubSub class) layer
* Added tests
* Added system-tests
* Added samples
* Added samples system-tests

### PR DESCRIPTION
**Pull request by feywind**
fixes: https://github.com/googleapis/nodejs-pubsub/issues/1220

The remainder of this PR is about adding schema sample snippets, as well as some preliminary TypeScript snippet support. There are still some linting issues with the generated samples, and I wanted to talk about the method used here before committing to it, so it's still marked as a work in progress.

### CODE REVIEW

File: samples/createTopicWithSchema.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+'use strict';
+
+// sample-metadata:
+//   title: Create Topic With Schema
+//   description: Creates a new topic, with a schema definition.
+//   usage: node createTopicWithSchema.js <topic-name>
+
+async function main(topicName = 'YOUR_TOPIC_NAME') {
+  // [START pubsub_create_topic]
```
Comment (anguillanneuf) [inline samples/createTopicWithSchema.js:31]: Is this sample supposed to create a topic with schema? If so, need to pass schema name and update the region tag:

```suggestion
  // [START pubsub_create_topic_with_schema]
```

File: samples/createTopicWithSchema.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  const pubSubClient = new PubSub();
+
+  async function createTopic() {
+    // Creates a new topic
+    await pubSubClient.createTopic(topicName);
+    console.log(`Topic ${topicName} created.`);
+  }
+
+  createTopic();
+  // [END pubsub_create_topic]
```
Comment (anguillanneuf) [inline samples/createTopicWithSchema.js:50]: Ditto

**Review by anguillanneuf — CHANGES_REQUESTED** [review]

File: samples/createTopicWithSchema.js (hunk 1 above)
Comment (feywind) [reply samples/createTopicWithSchema.js:31]: I actually hadn't even started on this one, but your change is right. Thanks :)

File: samples/listenForAvroRecords.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    let result;
+    switch (schemaMetadata.encoding) {
+      case Encodings.Binary:
+        result = type.fromBuffer(message.data);
+        break;
+      case Encodings.Json:
+        result = type.fromString(message.data.toString());
+        break;
+    }
```
Comment (anguillanneuf) [inline samples/listenForAvroRecords.js:78]: Should also handle an all-else case?

File: samples/listenForProtobufMessages.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      case Encodings.Binary:
+        result = Province.decode(message.data);
+        break;
+      case Encodings.Json:
+        result = JSON.parse(message.data.toString());
+        // What's coming in here is not properly protobuf data, but you could
+        // verify it if you like:
+        // assert.strictEqual(null, Province.verify(result));
+        break;
+    }
```
Comment (anguillanneuf) [inline samples/listenForProtobufMessages.js:79]: Ditto. Should also handle an all-else case?

File: samples/publishAvroRecords.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  };
+  let dataBuffer;
+  switch (schemaEncoding) {
+    case Encodings.Binary:
+      dataBuffer = type.toBuffer(province);
+      break;
+    case Encodings.Json:
+      dataBuffer = Buffer.from(type.toString(province));
+      break;
+  }
```
Comment (anguillanneuf) [inline samples/publishAvroRecords.js:79]: Ditto. `schemaEncoding` could be undefined, so an all-else case is helpful.

File: samples/publishProtobufMessages.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  let dataBuffer;
+  switch (schemaEncoding) {
+    case Encodings.Binary:
+      dataBuffer = Buffer.from(Province.encode(message).finish());
+      break;
+    case Encodings.Json:
+      dataBuffer = Buffer.from(JSON.stringify(message.toJSON()));
+      break;
+  }
```
Comment (anguillanneuf) [inline samples/publishProtobufMessages.js:81]: Ditto.

File: samples/typescript/listenForAvroRecords.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    let result: object | undefined;
+    switch (schemaMetadata.encoding) {
+      case Encodings.Binary:
+        result = type.fromBuffer(message.data);
+        break;
+      case Encodings.Json:
+        result = type.fromString(message.data.toString());
+        break;
+    }
```
Comment (anguillanneuf) [inline samples/typescript/listenForAvroRecords.ts:74]: ditto.

File: samples/typescript/listenForProtobufMessages.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      case Encodings.Binary:
+        result = Province.decode(message.data);
+        break;
+      case Encodings.Json:
+        result = JSON.parse(message.data.toString());
+        // What's coming in here is not properly protobuf data, but you could
+        // verify it if you like:
+        // assert.strictEqual(null, Province.verify(result));
+        break;
+    }
```
Comment (anguillanneuf) [inline samples/typescript/listenForProtobufMessages.ts:78]: ditto.

File: samples/typescript/publishAvroRecords.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  };
+  let dataBuffer: Buffer | undefined;
+  switch (schemaEncoding) {
+    case Encodings.Binary:
+      dataBuffer = type.toBuffer(province);
+      break;
+    case Encodings.Json:
+      dataBuffer = Buffer.from(type.toString(province));
+      break;
+  }
```
Comment (anguillanneuf) [inline samples/typescript/publishAvroRecords.ts:80]: ditto.

File: samples/typescript/publishProtobufMessages.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+  let dataBuffer: Buffer | undefined;
+  switch (schemaEncoding) {
+    case Encodings.Binary:
+      dataBuffer = Buffer.from(Province.encode(message).finish());
+      break;
+    case Encodings.Json:
+      dataBuffer = Buffer.from(JSON.stringify(message.toJSON()));
+      break;
+  }
```
Comment (anguillanneuf) [inline samples/typescript/publishProtobufMessages.ts:82]: ditto.

**Review by anguillanneuf — CHANGES_REQUESTED** [review]
See inline comments.

File: samples/listenForProtobufMessages.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    let result;
+    switch (schemaMetadata.encoding) {
+      case Encodings.Binary:
+        result = Province.decode(message.data);
+        break;
+      case Encodings.Json:
+        result = JSON.parse(message.data.toString());
+        // What's coming in here is not properly protobuf data, but you could
+        // verify it if you like:
+        // assert.strictEqual(null, Province.verify(result));
```
Comment (anguillanneuf) [inline samples/listenForProtobufMessages.js:77]: Could a Node.js folk comment here if this is necessary or slightly confusing? My understanding is that a protobuf message can be encoded in JSON or BINARY. The comment about "not properly protobuf data" is bit confusing.

File: samples/listenForAvroRecords.js (hunk 3 above)
Comment (feywind) [reply samples/listenForAvroRecords.js]: I went ahead and added default cases for all of these.

File: samples/listenForProtobufMessages.js (hunk 11 above)
Comment (feywind) [reply samples/listenForProtobufMessages.js]: I've just changed this to explicitly do the `verify()`. I think that should clear up the confusion.

**Review by anguillanneuf — APPROVED** [review]
LGTM to me overall. I'm not a Node.js expert. 😃

**Pull request merged**